src/lib.rs

```diff
@@ -2724,6 +2961,79 @@ impl Connection {
         Ok(())
     }
 
+    #[cfg(feature = "qlogging")]
+    fn qlog_recovery_metrics_updated(&mut self) {
```

Comment:
It seems in general these methods return the event, and then the caller writes them to the writer, so maybe this could be a `to_qlog()` or whatever method of `recovery::Recovery` that returns the event, and then we can use the `qlog_with!()` to avoid a lot of duplication.

(yes, this would require modifying the recovery module, but not the actual recovery code which was the main annoyance with the previous version, as this method would be self-contained and isolated from the rest of the code, also, this would allow accessing some additional internal values without exposing them outside of the module).